File: remove_node_from_end.py
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head   #creating a link
        p1,p2 =dummy,dummy
        for i in range(n+1):
            p2 = p2.next
            
        while(p2!=None):
            p1 = p1.next
            p2 = p2.next
            
        p1.next = p1.next.next
        return dummy.next
        
        # time complexity is o(n) n being number of nodes and space complexity is constant.
        #I am creating a dummy node and then pointing two pointers p1 and p2 at it, moving p2 pointer n nodes ahead of p1.
        #then while p2 is not null, i am traversing p1 and p2 ahead.
        #when p2 is null, p1 will be at previous node of the node which needs to be deleted and hence will point prev node to next node hence will delete the required node.

File: RemoveNodeFromEnd.java
// Time Complexity :O(N) where n is the number of nodes 
// Space Complexity :O(1) 
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach: maintaing two pointers slow and fast where we start iterating the slow 
// after fast has moved n nodes.When fast reaches null we, move the next pointer of slow  to the next next pointer, thus skipping the
// desired the node.
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head==null || head.next==null)
            return null;
        ListNode dummy = new ListNode(-1);
        ListNode slow =dummy, fast = dummy;
        dummy.next=head;int i=0;
        while(i<=n){
            fast=fast.next;
            i++;
        }
        while(fast!=null){
            slow=slow.next;
            fast=fast.next;
        }
        slow.next =slow.next.next;
        
    return dummy.next;
    }
}